enterprise/server/remote_execution/uffd/uffd.go

```diff
@@ -267,8 +300,16 @@ func pageStartAddress(addr uint64, pageSize int) uintptr {
 	return uintptr(addr & ^(uint64(pageSize) - 1))
 }
 
-func (h *Handler) Stop() {
-	h.lis.Close()
+func (h *Handler) Stop() error {
+	log.Info("UFFD handler beginning shut down")
+	_, err := h.earlyTerminationWriter.Write([]byte{0})
+	if err != nil {
+		return status.WrapError(err, "write to early terminator")
+	}
+	h.wg.Wait()
+	h.earlyTerminationReader.Close()
+	h.earlyTerminationWriter.Close()
```

Comment:
sorry missed this - I think the `h.wg.Wait()` needs to go after the Close() here, since the Close() is what awakens the uffd poll?

Reply:
h.earlyTerminationWriter.Write() is what awakens the uffd poll.

io.Pipe.Write() writes to the pipe immediately - it does not wait for Close() to flush the data. I put the h.wg.Wait() after because I didn't want to close the earlyTerminationReader before it could read the early termination message